File: mobile/test/unit/hotel_time_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:marina_hotel_mobile/utils/hotel_time_engine.dart';

void main() {
  group('HotelTimeEngine', () {
    group('getHotelDayKey', () {
      test('should return previous day before 14:01', () {
        // 10:00 AM on July 19 → hotel day = July 18
        final dt = DateTime(2026, 7, 19, 10, 0);
        final key = HotelTimeEngine.getHotelDayKey(dateTime: dt);
        expect(key, '2026-07-18');
      });

      test('should return current day at exactly 14:01', () {
        // 14:01 on July 19 → hotel day = July 19
        final dt = DateTime(2026, 7, 19, 14, 1);
        final key = HotelTimeEngine.getHotelDayKey(dateTime: dt);
        expect(key, '2026-07-19');
      });

      test('should return previous day at 14:00', () {
        // 14:00 on July 19 → hotel day = July 18 (still in previous hotel day)
        final dt = DateTime(2026, 7, 19, 14, 0);
        final key = HotelTimeEngine.getHotelDayKey(dateTime: dt);
        expect(key, '2026-07-18');
      });

      test('should return current day after 14:01', () {
        // 15:00 on July 19 → hotel day = July 19
        final dt = DateTime(2026, 7, 19, 15, 0);
        final key = HotelTimeEngine.getHotelDayKey(dateTime: dt);
        expect(key, '2026-07-19');
      });

      test('should handle midnight correctly', () {
        // 00:00 on July 19 → hotel day = July 18
        final dt = DateTime(2026, 7, 19, 0, 0);
        final key = HotelTimeEngine.getHotelDayKey(dateTime: dt);
        expect(key, '2026-07-18');
      });

      test('should handle 23:59 correctly', () {
        // 23:59 on July 19 → hotel day = July 19
        final dt = DateTime(2026, 7, 19, 23, 59);
        final key = HotelTimeEngine.getHotelDayKey(dateTime: dt);
        expect(key, '2026-07-19');
      });
    });

    group('calculateDays', () {
      test('same day should be 1 night minimum', () {
        final checkin = DateTime(2026, 7, 19, 14, 1);
        final checkout = DateTime(2026, 7, 19, 18, 0);
        final days = HotelTimeEngine.calculateDays(checkin, checkOut: checkout);
        expect(days, 1);
      });

      test('2 days checkout before 14:01 = 1 night', () {
        final checkin = DateTime(2026, 7, 19, 14, 1);
        final checkout = DateTime(2026, 7, 20, 14, 0);
        final days = HotelTimeEngine.calculateDays(checkin, checkOut: checkout);
        expect(days, 1);
      });

      test('2 days checkout at 14:01 = 2 nights', () {
        final checkin = DateTime(2026, 7, 19, 14, 1);
        final checkout = DateTime(2026, 7, 20, 14, 1);
        final days = HotelTimeEngine.calculateDays(checkin, checkOut: checkout);
        expect(days, 2);
      });

      test('3 days checkout at 14:00 = 2 nights', () {
        final checkin = DateTime(2026, 7, 19, 14, 1);
        final checkout = DateTime(2026, 7, 21, 14, 0);
        final days = HotelTimeEngine.calculateDays(checkin, checkOut: checkout);
        expect(days, 2);
      });

      test('3 days checkout at 14:01 = 3 nights', () {
        final checkin = DateTime(2026, 7, 19, 14, 1);
        final checkout = DateTime(2026, 7, 21, 14, 1);
        final days = HotelTimeEngine.calculateDays(checkin, checkOut: checkout);
        expect(days, 3);
      });
    });
  });
}

File: mobile/android/app/src/test/java/com/marina/marina/domain/util/HotelTimeEngineTest.kt
package com.marina.marina.domain.util

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Calendar

/**
 * HotelTimeEngine unit tests — the financial day-boundary contract
 * (14:01 hotel day) ported from the Flutter `hotel_time_engine_test.dart`
 * expectations.
 */
class HotelTimeEngineTest {

    private fun millis(year: Int, month: Int, day: Int, hour: Int, minute: Int, second: Int = 0): Long {
        val cal = Calendar.getInstance()
        cal.set(year, month - 1, day, hour, minute, second)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    // -------------------------------------------------------------------------
    // Hotel day key resolution
    // -------------------------------------------------------------------------

    @Test
    fun `before 1401 belongs to previous hotel day`() {
        // 2026-09-18 14:00:59 → hotel day 2026-09-17
        val key = HotelTimeEngine.hotelDayKey(millis(2026, 9, 18, 14, 0, 59))
        assertEquals("2026-09-17", key)
    }

    @Test
    fun `exactly 1401 starts the current hotel day`() {
        val key = HotelTimeEngine.hotelDayKey(millis(2026, 9, 18, 14, 1, 0))
        assertEquals("2026-09-18", key)
    }

    @Test
    fun `after 1401 belongs to current hotel day`() {
        val key = HotelTimeEngine.hotelDayKey(millis(2026, 9, 18, 23, 30, 0))
        assertEquals("2026-09-18", key)
    }

    @Test
    fun `early morning belongs to previous calendar day`() {
        // 2026-09-18 01:00 → hotel day 2026-09-17
        val key = HotelTimeEngine.hotelDayKey(millis(2026, 9, 18, 1, 0, 0))
        assertEquals("2026-09-17", key)
    }

    @Test
    fun `noon belongs to previous calendar day`() {
        val key = HotelTimeEngine.hotelDayKey(millis(2026, 9, 18, 12, 0, 0))
        assertEquals("2026-09-17", key)
    }

    // -------------------------------------------------------------------------
    // Cutoff helper
    // -------------------------------------------------------------------------

    @Test
    fun `cutoff boundaries`() {
        assertFalse(HotelTimeEngine.isAfterCutoff(millis(2026, 9, 18, 13, 59, 59)))
        assertTrue(HotelTimeEngine.isAfterCutoff(millis(2026, 9, 18, 14, 1, 0)))
        assertTrue(HotelTimeEngine.isAfterCutoff(millis(2026, 9, 18, 15, 0, 0)))
        assertFalse(HotelTimeEngine.isAfterCutoff(millis(2026, 9, 18, 0, 0, 0)))
    }

    // -------------------------------------------------------------------------
    // Day count / pricing
    // -------------------------------------------------------------------------

    @Test
    fun `one night when checkout next day before 1401`() {
        // Check-in 2026-09-01 14:01 → Check-out 2026-09-02 14:00 = 1 day
        val days = HotelTimeEngine.calculateDays(
            millis(2026, 9, 1, 14, 1, 0),
            millis(2026, 9, 2, 14, 0, 0)
        )
        assertEquals(1, days)
    }

    @Test
    fun `two nights when checkout at or after 1401`() {
        // Check-in 2026-09-01 14:01 → Check-out 2026-09-02 14:01 = 2 days
        val days = HotelTimeEngine.calculateDays(
            millis(2026, 9, 1, 14, 1, 0),
            millis(2026, 9, 2, 14, 1, 0)
        )
        assertEquals(2, days)
    }

    @Test
    fun `same calendar dates count as at least one day`() {
        // Check-in 01:00 → Check-out 10:00 same date (before the 14:01 cutoff):
        // calendar diff 0 → minimum 1 day, no cutoff surcharge = 1 day.
        val days = HotelTimeEngine.calculateDays(
            millis(2026, 9, 1, 1, 0, 0),
            millis(2026, 9, 1, 10, 0, 0)
        )
        assertEquals(1, days)
    }

    @Test
    fun `same-day checkout after cutoff bills two days`() {
        // Flutter contract parity: calendar diff 0 → 1, checkout 20:00 is at
        // or after 14:01 → +1 surcharge = 2 days (identical to Dart engine).
        val days = HotelTimeEngine.calculateDays(
            millis(2026, 9, 1, 15, 0, 0),
            millis(2026, 9, 1, 20, 0, 0)
        )
        assertEquals(2, days)
    }

    @Test
    fun `total amount multiplies days by nightly price`() {
        val total = HotelTimeEngine.calculateTotalAmount(
            pricePerNight = 250.0,
            checkIn = millis(2026, 9, 1, 14, 1, 0),
            checkOut = millis(2026, 9, 3, 14, 0, 0)
        )
        assertEquals(500.0, total, 0.001)
    }

    @Test
    fun `discount subtracted and clamped to zero`() {
        val total = HotelTimeEngine.calculateTotalAmount(
            pricePerNight = 200.0,
            checkIn = millis(2026, 9, 1, 14, 1, 0),
            checkOut = millis(2026, 9, 2, 14, 0, 0),
            discount = 300.0
        )
        assertEquals(0.0, total, 0.001)
    }

    // -------------------------------------------------------------------------
    // Interval helpers
    // -------------------------------------------------------------------------

    @Test
    fun `hotel day start and end bracket the day`() {
        val ref = millis(2026, 9, 18, 20, 0, 0)
        val start = HotelTimeEngine.hotelDayStart(ref)
        val end = HotelTimeEngine.hotelDayEnd(ref)

        val startCal = Calendar.getInstance().apply { timeInMillis = start }
        assertEquals(14, startCal.get(Calendar.HOUR_OF_DAY))
        assertEquals(1, startCal.get(Calendar.MINUTE))
        assertEquals(18, startCal.get(Calendar.DAY_OF_MONTH))

        assertEquals(24 * 60 * 60 * 1000L, end - start)
    }

    // -------------------------------------------------------------------------
    // Parsing
    // -------------------------------------------------------------------------

    @Test
    fun `parses iso and display formats`() {
        assertTrue(HotelTimeEngine.parseDate("2026-09-18T20:15:00") != null)
        assertTrue(HotelTimeEngine.parseDate("2026-09-18 20:15") != null)
        assertTrue(HotelTimeEngine.parseDate("18/09/2026 20:15") != null)
        assertNull(HotelTimeEngine.parseDate(null))
        assertNull(HotelTimeEngine.parseDate(""))
    }
}
